--- a/NANOVNA_STM32_F303/adc.c
+++ b/NANOVNA_STM32_F303/adc.c
@@ -86,8 +86,11 @@ static ADCConversionGroup adcgrpcfgXY = {
   {ADC_SQR1_SQ1_N(ADC_CHANNEL_IN3), 0, 0, 0} /* SQR[4]  */
 };
 
+static mutex_t adc_mutex;
+
 void adc_init(void)
 {
+  osalMutexObjectInit(&adc_mutex);
   adcStart(&ADCD2, NULL);
   adcStart(&ADCD1, NULL);
   #ifdef F303_ADC_VREF_ALWAYS_ON
@@ -102,7 +105,9 @@ uint16_t adc_single_read(uint32_t chsel)
   /* ADC setup */
 //  adcStart(&ADCD2, NULL);
   adcgrpcfgXY.sqr[0] = ADC_SQR1_SQ1_N(chsel);
+  osalMutexLock(&adc_mutex);
   adcConvert(&ADCD2, &adcgrpcfgXY, samples, 1);
+  osalMutexUnlock(&adc_mutex);
   return(samples[0]);
 }
 
@@ -110,7 +115,9 @@ uint16_t adc1_single_read(uint32_t chsel)
 {
   (void)chsel;
   /* ADC setup */
+  osalMutexLock(&adc_mutex);
   adcConvert(&ADCD1, &adcgrpcfgVersion, samples, 1);
+  osalMutexUnlock(&adc_mutex);
   return(samples[0]);
 }
 
@@ -144,6 +151,7 @@ int16_t adc_vbat_read(void)
 //                               Vref+ = 3.3 V (tolerance: +-10 mV). */
 //  float avg_slope = ((float)(temperature_cal1 - temperature_cal2))/(110-25);
 //  float ts;
+  osalMutexLock(&adc_mutex);
 #ifndef F303_ADC_VREF_ALWAYS_ON
   adcSTM32EnableVBAT(&ADCD1);
   adcSTM32EnableVREF(&ADCD1);
@@ -161,6 +169,7 @@ int16_t adc_vbat_read(void)
   adcSTM32DisableVREF(&ADCD1);
 //  adcSTM32DisableTS(&ADCD1);
 #endif
+  osalMutexUnlock(&adc_mutex);
 
 //  ts = samplesVBAT[2];
 //  uint16_t vts = (ADC_FULL_SCALE * VREFINT_CAL * ts / (vrefint * ((1<<12)-1)));
diff --git a/chconf.h b/chconf.h
index 682a459..d11185e 100644
--- a/chconf.h
+++ b/chconf.h
@@ -187,7 +187,7 @@
  *
  * @note    The default is @p TRUE.
  */
-#define CH_CFG_USE_MUTEXES                  FALSE
+#define CH_CFG_USE_MUTEXES                  TRUE
 
 /**
  * @brief   Enables recursive behavior on mutexes.
